File: internal/dsl/lexer/reader.go
package lexer

import (
	"strings"
	"unicode"

	"github.com/infraflakes/sro/internal/dsl/token"
)

func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\r' || l.ch == '\n' {
		l.readChar()
	}
}

func (l *Lexer) skipComment() {
	if l.ch != '#' {
		return
	}
	for l.ch != '\n' && l.ch != 0 {
		l.readChar()
	}
}

func (l *Lexer) peek() byte {
	if l.readPos >= len(l.input) {
		return 0
	}
	return l.input[l.readPos]
}

func (l *Lexer) readBacktick() token.Token {
	line := l.line
	col := l.col
	l.readChar() // consume opening `

	var lit strings.Builder
	for l.ch != '`' && l.ch != 0 {
		lit.WriteByte(l.ch)
		l.readChar()
	}

	if l.ch != '`' {
		return token.Token{
			Type:    token.ILLEGAL,
			Literal: "unterminated backtick string",
			Line:    line,
			Col:     col,
		}
	}

	return token.Token{
		Type:    token.BACKTICK,
		Literal: lit.String(),
		Line:    line,
		Col:     col,
	}
}

func (l *Lexer) readPath() token.Token {
	line := l.line
	col := l.col
	l.readChar() // consume '.'
	l.readChar() // consume '/'

	var lit strings.Builder
	lit.WriteString("./")

	for l.ch != 0 && l.ch != ' ' && l.ch != '\t' && l.ch != '\n' && l.ch != ',' && l.ch != ']' && l.ch != ';' {
		lit.WriteByte(l.ch)
		l.readChar()
	}

	return token.Token{
		Type:    token.PATH_LIT,
		Literal: lit.String(),
		Line:    line,
		Col:     col,
	}
}

func (l *Lexer) readIdent() token.Token {
	line := l.line
	col := l.col

	var lit strings.Builder
	for l.ch != 0 && (unicode.IsLetter(rune(l.ch)) || l.ch == '_' || l.ch == '-' || unicode.IsDigit(rune(l.ch))) {
		lit.WriteByte(l.ch)
		l.readChar()
	}

	return token.Token{
		Type:    token.LookupIdent(lit.String()),
		Literal: lit.String(),
		Line:    line,
		Col:     col,
	}
}

File: src/dsl/lexer/reader.rs
use crate::dsl::token::{Token, TokenType, lookup_ident};
use super::Lexer;

impl Lexer {
    pub(super) fn read_char(&mut self) {
        if let Some(c) = self.ch {
            self.byte_offset += c.len_utf8();
        }
        self.ch = if self.read_pos < self.input.len() {
            Some(self.input[self.read_pos])
        } else {
            None
        };
        self.pos = self.read_pos;
        self.read_pos += 1;

        if self.ch == Some('\n') {
            self.line += 1;
            self.col = 0;
        } else {
            self.col += 1;
        }
    }

    pub(super) fn peek(&self) -> Option<char> {
        if self.read_pos < self.input.len() {
            Some(self.input[self.read_pos])
        } else {
            None
        }
    }

    pub(super) fn skip_whitespace(&mut self) {
        while let Some(c) = self.ch {
            if !c.is_whitespace() {
                break;
            }
            self.read_char();
        }
    }

    pub(super) fn skip_comment(&mut self) {
        while self.ch != Some('\n') && self.ch.is_some() {
            self.read_char();
        }
    }

    pub(super) fn read_ident(&mut self) -> Token {
        let start_line = self.line;
        let start_col = self.col;
        let start_pos = self.pos;
        let start_byte_offset = self.byte_offset;

        while let Some(c) = self.ch {
            if c.is_alphanumeric() || c == '_' {
                self.read_char();
            } else {
                break;
            }
        }

        let ident: String = self.input[start_pos..self.pos].iter().collect();
        let ty = lookup_ident(&ident);
        Token::new(
            ty,
            start_line,
            start_col,
            start_byte_offset,
            self.byte_offset - start_byte_offset,
        )
    }

    pub(super) fn read_backtick(&mut self) -> Token {
        let start_line = self.line;
        let start_col = self.col;
        let start_pos = self.pos;
        let start_byte_offset = self.byte_offset;

        self.read_char(); // skip opening backtick
        while let Some(c) = self.ch {
            if c == '`' {
                break;
            }
            if c == '\n' {
                // Unterminated backtick string - stop at newline
                break;
            }
            self.read_char();
        }

        let content: String = self.input[start_pos + 1..self.pos].iter().collect();

        if self.ch == Some('`') {
            self.read_char(); // skip closing backtick
            Token::new(
                TokenType::Backtick(content),
                start_line,
                start_col,
                start_byte_offset,
                self.byte_offset - start_byte_offset,
            )
        } else {
            // Unterminated backtick string
            Token::new(
                TokenType::Illegal("unterminated backtick string".to_string()),
                start_line,
                start_col,
                start_byte_offset,
                self.byte_offset - start_byte_offset,
            )
        }
    }

    pub(super) fn read_path(&mut self) -> Token {
        let start_line = self.line;
        let start_col = self.col;
        let start_pos = self.pos;
        let start_byte_offset = self.byte_offset;

        self.read_char(); // skip '.'
        self.read_char(); // skip '/'

        while let Some(c) = self.ch {
            if !c.is_whitespace() && c != ',' && c != ']' && c != ';' {
                self.read_char();
            } else {
                break;
            }
        }

        let path: String = self.input[start_pos..self.pos].iter().collect();
        Token::new(
            TokenType::PathLit(path),
            start_line,
            start_col,
            start_byte_offset,
            self.byte_offset - start_byte_offset,
        )
    }
}
